### ISSUE TITLE
curl needs update to latest

### ISSUE THREAD

**Issue by mwoehlke-kitware** [RobotLocomotion/drake#18032]
curl needs to be updated to the latest release.

It appears the definitions have changed and need to be updated. The build fails with "Missing define or undefine decision for HAVE_ATOMIC when running in strict=True mode".

**Comment by mwoehlke-kitware** [comment]
BTW, I suspect this isn't difficult to resolve, but "how" wasn't immediately clear from the `repository.bzl` comment. I wonder if making that just a little more detailed would make it easier for someone not otherwise familiar with how Drake bundles curl to get past such bumps.

**Comment by jwnimmer-tri** [comment]
(1) I ran `new_release` to upgrade the git tag, and then ran a build.

(2) The error message from bazel told me that I need to specify `HAVE_ATOMIC`.

Grepping, the curl source code explains the flag like so:

https://github.com/curl/curl/blob/93d092867f0f2c78571983040ef75e078ee1a4c4/lib/curl_config.h.cmake#L159-L160

https://github.com/curl/curl/blob/93d092867f0f2c78571983040ef75e078ee1a4c4/CMake/CurlTests.c#L519-L528

I looked at both Ubuntu and macOS and observed that they _do_ both have `<stdatomic.h>` so added it to the "defined for all of our platforms" list in `cmakedefines.bzl`.

If the flag were some skeevy feature like `HAVE_LDAP_H`, I would have disabled it even if both platforms happened to it.  But `stdatomic.h` seems like a good idea to use.

(3) The error message from bazel told me to delete a bunch of unused definitions, so I did that.
